utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -328,10 +328,110 @@ where
 	Ok(())
 }
 
+async fn execute_block<Block, ExecDispatch>(
+	shared: SharedParams,
+	command: ExecuteBlockCmd,
+	config: Configuration,
+)-> sc_cli::Result<()>
+where
+	Block: BlockT + serde::de::DeserializeOwned,
+	Block::Hash: FromStr,
+	<Block::Hash as FromStr>::Err: Debug,
+	NumberFor<Block>: FromStr,
+	<NumberFor<Block> as FromStr>::Err: Debug,
+	ExecDispatch: NativeExecutionDispatch + 'static,
+{
+	let wasm_method = shared.wasm_method;
+	let execution = shared.execution;
+	let heap_pages = shared.heap_pages.or(config.default_heap_pages);
+
+	let mut changes = Default::default();
+	let max_runtime_instances = config.max_runtime_instances;
+	let executor = NativeExecutor::<ExecDispatch>::new(
+		wasm_method.into(),
+		heap_pages,
+		max_runtime_instances,
+	);
+
+	let (mode, block) = match command.state {
+		State::Snap{ snapshot_path, url } => {
+			let header_hash: Block::Hash = shared.block_at
+				.parse()
+				.map_err(|e| format!("Could not parse header hash: {:?}", e))?;
+			let block: Block = rpc_api::get_block::<Block, _>(url.clone(), header_hash).await?;
+
+			let mode = Mode::Offline(OfflineConfig {
+				state_snapshot: SnapshotConfig::new(snapshot_path),
+			});
+
+			(mode, block)
+		},
+		State::Live { url, snapshot_path, modules } => {
+			let block_hash: Block::Hash = shared.block_at
+				.parse()
+				.map_err(|e| format!("Could not parse hash: {:?}", e))?;
+
+			let block: Block = rpc_api::get_block::<Block, _>(url.clone(), block_hash).await?;
```

Comment:
Looks like both branches of this `match` get the block in identical ways. This suggests that you could extract the block-getting machinery to before the `match`, so that it only has to compute `mode`.